File: python/src/base/drivers/base_web_driver.py
import os

from datetime import datetime
from selenium import webdriver
from selenium.webdriver.chrome.webdriver import WebDriver
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.remote.webelement import WebElement
from selenium.webdriver.support.select import Select
import time
from common.file import CommonFile
from webdriver_manager.chrome import ChromeDriverManager

class BaseWebDriver:
    def __init__(cls):
        pass
    
    driver: WebDriver = None
    
    @classmethod
    def start_driver(cls):
        options = webdriver.ChromeOptions()
        options.add_argument('--lang=ja-JP')
        options.add_experimental_option('detach', True)
        cls.driver = webdriver.Chrome(ChromeDriverManager().install(),options=options)
        cls.driver.maximize_window()

    @classmethod
    def stop_driver(cls):
        cls.driver.quit()
    
    @classmethod
    def log_current_url(cls):
        print('current URL: ', cls.driver.current_url)

    @classmethod
    def get_screenshot(cls, id: str = "temp", 
                       prefix: str = "", suffix: str = ""):
        now = datetime.now()
        prefix = f"{prefix}_" if prefix else ""
        suffix = f"_{suffix}" if suffix else ""
        file_name = f"{id}_{prefix}{now.strftime('%Y%m%d%H%M%S%f')}{suffix}.png"
        dir_path = f"{os.getcwd()}/screenshots/"
        file_path = f"{dir_path}{file_name}"
        
        if not CommonFile.exists_path(dir_path):
            CommonFile.make_directory(dir_path)

        print(f'screenshot:{file_path}')
        screenshot = cls.driver.get_screenshot_as_png()
    
        CommonFile.make(file_path, screenshot, mode='wb+')

    @classmethod
    def switch_to_window(cls, index=-1):
        # seleniumが速すぎるため少し待つ
        cls.wait(0.5)
        
        windows = cls.driver.window_handles
        if index == -1:
            index = len(windows) - 1
        cls.driver.switch_to.window(windows[index])

    @classmethod
    def find_wait_clickable_element(cls, selector, wait_seconds=5):
        wait = WebDriverWait(cls.driver, wait_seconds)
        elm = wait.until(EC.element_to_be_clickable(selector))
        return elm
    
    @classmethod
    def find_wait_located_element(cls, selector, wait_seconds=10):
        wait = WebDriverWait(cls.driver, wait_seconds)
        elm = wait.until(EC.visibility_of_element_located(selector))
        return elm

    @classmethod
    def wait_loading(cls, selector, wait_seconds=10):
        # 判定用のエレメントが読込されるまで待機する
        wait = WebDriverWait(cls.driver, wait_seconds)
        elm = wait.until(EC.visibility_of_element_located(selector))
    
    @classmethod
    def wait(cls, wait_seconds=1):
        time.sleep(wait_seconds)
    
    @classmethod
    def clear_text(cls, elm: WebElement):
        elm.clear()
        
    @classmethod
    def input_text(cls, elm: WebElement, value):
        cls.clear_text(elm)
        elm.send_keys(value)
    
    @classmethod
    def input_text_add(cls, elm: WebElement, value):
        elm.send_keys(value)
        
    @classmethod
    def input_checkbox(cls, elm: WebElement, is_check: bool):
        is_checked = elm.is_selected()
        if is_checked == is_check:
            return
        elm.click()

    @classmethod
    def input_select(cls, elm: WebElement, value):
        Select(elm).select_by_value(value)
    
    @classmethod
    def input_radio(cls, name: str , value):
        elm_list = cls.driver.find_elements(by=By.NAME, value=name)
        elm = next(filter(lambda x:x.get_attribute('value') == value, elm_list))
        elm.click()
        
    @classmethod
    def select_radio(cls, elm: WebElement):
        elm.click()
    

File: csharp/SeleniumSample/Src/Base/Drivers/BaseWebDriver.cs

using System.Diagnostics;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Support.UI;
using System.Threading;
using SeleniumExtras.WaitHelpers;
using Common;

using WebDriverManager;
using WebDriverManager.DriverConfigs.Impl;
using WebDriverManager.Helpers;

namespace Base.Drivers
{
    class BaseWebDriver
    {

        // コンストラクタを定義します。
        public BaseWebDriver()
        {
            this.driver = this.StartDriver();
        }

        public ChromeDriver driver;

        public ChromeDriver StartDriver()
        {
            // TODO:自動更新のためのコード。ただしダウンロード先が実行フォルダ直下になるためコメントアウト中
            // new DriverManager().SetUpDriver(new ChromeConfig(), VersionResolveStrategy.MatchingBrowser);

            var options = new ChromeOptions();
            options.AddArguments("--lang=ja-JP");
            // options.AddAdditionalChromeOption("detach", true);
            ChromeDriver driver = new ChromeDriver(options);
            driver.Manage().Window.Maximize();
            return driver;
        }

        public void StopDriver()
        {
            this.driver.Quit();
        }

        public void LogCurrentUrl()
        {
            Debug.WriteLine("current URL: " + this.driver.Url);
        }

        public void GetScreenshot(string id = "temp", string prefix = "", string suffix = "")
        {
            var now = DateTime.Today;
            var timestamp = now.ToString("yyyyMMddHHmmssfff");
            prefix = !string.IsNullOrEmpty(prefix) ? $"{prefix}_" : "";
            suffix = !string.IsNullOrEmpty(prefix) ? $"_{suffix}" : "";
            var fileName = $"{id}_{prefix}{timestamp}{suffix}.png";
            var dirPath = $"{CommonFile.GetCwd()}/screenshots/";
            var filePath = $"{dirPath}${fileName}";

            if (!CommonFile.ExistsPath(dirPath)){
                CommonFile.CreateDirectory(dirPath);
            }

            Console.WriteLine($"screenshot:{filePath}");
            Screenshot screenshot = (driver as ITakesScreenshot).GetScreenshot();
            screenshot.SaveAsFile(filePath, ScreenshotImageFormat.Png);
        }


        public void SwitchToWindow(int index)
        {
            var windows = this.driver.WindowHandles;
            if (index == -1)
            {
                index = windows.Count - 1;
            }
            this.driver.SwitchTo().Window(windows[index]);
        }

        public IWebElement FindWaitClickableElement(By selector, double waitSeconds = 10)
        {
            WebDriverWait wait = new WebDriverWait(driver, TimeSpan.FromSeconds(waitSeconds));
            IWebElement firstResult = wait.Until(ExpectedConditions.ElementToBeClickable((selector)));
            IWebElement elm = (IWebElement)firstResult;
            return elm;
        }

        public IWebElement FindWaitLocatedElement(By selector, double waitSeconds = 10)
        {
            WebDriverWait wait = new WebDriverWait(driver, TimeSpan.FromSeconds(waitSeconds));
            var firstResult = wait.Until(ExpectedConditions.ElementExists((selector)));
            IWebElement elm = (IWebElement)firstResult;
            return elm;
        }

        public void WaitLoading(By selector, double waitSeconds = 10)
        {
            // 判定用のエレメントが読込されるまで待機する
            WebDriverWait wait = new WebDriverWait(driver, TimeSpan.FromSeconds(waitSeconds));
            IWebElement firstResult = wait.Until(e => e.FindElement(selector));
        }

        public void Wait(double waitSeconds = 1)
        {
            Thread.Sleep((int)(waitSeconds * 1000));
        }

        public void ClearText(IWebElement elm)
        {
            elm.Clear();
        }

        public void InputText(IWebElement elm, string? value)
        {
            this.ClearText(elm);
            elm.SendKeys(value);
        }

        public void InputTextAdd(IWebElement elm, string? value)
        {
            elm.SendKeys(value);
        }

        public void InputCheckbox(IWebElement elm, bool isCheck)
        {
            var isChecked = elm.Selected;
            if (isChecked == isCheck)
            {
                return;
            }
            elm.Click();
        }

        public void InputSelect(IWebElement elm, string? value)
        {
            var select = new SelectElement(elm);
            select.SelectByValue(value);
        }

        public void InputRadio(List<IWebElement> elmList, string? value)
        {
            IWebElement elm;
            foreach (IWebElement x in elmList)
            {
                if (x.GetAttribute("value") == value)
                {
                    elm = x;
                    elm.Click();
                }
            }
        }
        public void InputRadio(string name, string? value)
        {
            var elmList = this.driver.FindElements(By.Name(name)).ToList<IWebElement>();
            this.InputRadio(elmList, value);
        }

        public void SelectRadio(IWebElement elm)
        {
            elm.Click();
        }

    }
}